### ISSUE TITLE
[Bug] iPhone.CopyFileFromDevice() fails and returns false

### ISSUE THREAD

**Issue by freemanrepo** [0xFireball/MK.MobileDevice#8]
iPhone.CopyFileFromDevice() fails and returns false while iPhone.CopyToFromDevice() works.

Tested on iOS 9.3.3

**Comment by 0xFireball** [comment]
Hello, sounds interesting. Can you please give more information about the error? I'll look at the code and ask for specifics soon.

**Comment by 0xFireball** [comment]
Which iPhone class are you using? May I see some of your error code?

**Comment by 0xFireball** [comment]
Also, which file are you copying? Do you know the permissions of that file? If you have a jailbroken device, you can easily check. If not, try copying a file to the device and copying it back and see if that works... If putting a file there works but pulling it back doesn't, it would probably be an issue with that specific file.

It would be helpful if you could give more information, possibly your code where you attempt to copy a file from the device.

Thanks. @freemanrepo

**Comment by freemanrepo** [comment]
I'm using the ```iPhone``` class.

What I'm doing is connecting to the device via house arrest to access an app container, then use the following:

```
string remotePath = "Documents/file.plist";
string localPath = System.IO.Path.GetTempPath() + System.IO.Path.GetFileName(remotePath);
iph.CopyFileFromDevice(localPath, remotePath);
```

That's enough to make it crash.
File ownership: ```mobile:mobile```, perms: ```rw-r-r```.

What's weird is that I haven't changed anything in my code and it used to work perfectly fine in the past, perhaps an iTunes update broke it?

**Comment by 0xFireball** [comment]
Oh, House Arrest. I know that Apple has been messing with that since iOS 8.3. They really locked it down since then. Before 8.3, you could HouseArrest connect to just about any app and edit data. Since then, I've been having serious issues with HouseArrest.

As for the internals of House arrest as used by `MK.MobileDevice`, it connects to HouseArrest and sends a `VendDocuments` command, then replaces the AFC handle (`hAFC`) with the one provided by HouseArrest.

```
        public bool ConnectViaHouseArrest(string appId)
        {
            bool ret;
            IntPtr currDevice = this.iPhoneHandle;
            IntPtr ldService;
            IntPtr lockdownClient;
            Lockdown.LockdownError lockdownReturnCode = Lockdown.Start(currDevice, out lockdownClient, out ldService);
            IntPtr hHA;
            IntPtr hSvc;
            HouseArrest.house_arrest_error_t hea = HouseArrest.house_arrest_client_start_service(currDevice, out hHA, out hSvc);
            hea = HouseArrest.house_arrest_send_command(hHA, "VendDocuments", appId);
            IntPtr result;
            hea = HouseArrest.house_arrest_get_result(hHA, out result);
            XDocument har = LibiMobileDevice.PlistToXml(result);
            PList pl = new PList(har, true);
            if (pl.ContainsKey("Error"))
            {
                ret = false;
            }
            else
            {
                IntPtr HA_afc;
                HouseArrest.afc_error_t afce = HouseArrest.afc_client_new_from_house_arrest_client(hHA, out HA_afc);
                ret = afce == HouseArrest.afc_error_t.AFC_E_SUCCESS;
                if (ret)
                    this.hAFC = HA_afc;
            }
            //Don't free the HA client because doing so makes the new AFC client fail.
            //HouseArrest.house_arrest_client_free(hHA);
            Lockdown.FreeService(ldService);
            Lockdown.FreeClient(lockdownClient);
            return ret;
        }
```

The issue seems to be that the `VendDocuments` command is failing. Ultimately, this is controlled by Apple, so I am limited in what I can do. Maybe you could try debugging the `ConnectViaHouseArrest` call to see what is going on underneath.

Oddly enough, you did say that copying files to the house arrest seemed to work. Maybe check the permissions on that file?

Do you control the iOS app? Maybe try changing the permissions (it doesn't look like it would work here, but weird stuff happens) Also, here's another suggestion that you might want to try; please note that the `iTMDiPhone` class for the most part uses `iTunesMobileDevice.dll`, while the `iPhone` class relies on the LibIMobileDevice libraries. This wiki page might help you: https://github.com/0xFireball/MK.MobileDevice/wiki/Getting-Started

When you say it worked fine in the past, do you mean with MK.MobileDevice?

I haven't updated it much since about April 2016, as I have been quite busy. I should have more time in about a month or so to get an iDevice and do some more testing. If you're really familiar with this stuff, I'd be happy to take pull requests.

**Comment by freemanrepo** [comment]
Yep, it's the exact same setup I've had in the past. MK.MobileDevice + same iOS version + same code used to work fine. Now it is no longer able to copy files from the device using the exact same setup, the only thing that was changed is I updated to the latest iTunes version.

I'll do more debugging and see what I can come up with.

**Comment by 0xFireball** [comment]
All right, let me know your results. As for debugging try not only .NET debugging but native debugging as well for the underlying native libraries.

**Comment by SpyderRoll** [comment]
I can confirm the same behavior on iOS 9.3.2, 9.3.4 and 9.3.5 -- CopyFileFromDevice() fails, but CopyFileToDevice() works OK.

**Comment by 0xFireball** [comment]
Weird. Have you tested it on a version below iOS 9.3? @freemanrepo says it used to work before.

At this point, it looks like it's beyond our control. Looks like it's Apple's fault :/

**Comment by SpyderRoll** [comment]
CopyFileFromDevice() does not work either when I tested it on iOS 8.1.2. I don't have iTunes installed on my PC, but I did recently install iTools, iExplorer and iMazing, which I believe downloaded and installed the latest Apple Application and Mobile Support files. This appears to be causing the issue. However, one thing to note, iTools copy works ok (with iOS version 9.3.x) -- no issues. I have not had a chance to test with iExplorer and iMazing, but I imagine it works as well.

**Comment by 0xFireball** [comment]
@SpyderRoll thanks for the observations. Actually I have noticed that iTools tries to register it's own (likely modified) DLL as the preferred iTunesMobileDevice library. Maybe you could debug MK.MobileDevice and find the DLL path?
Odd that iTools still works though. I'll look into that.

**Comment by 0xFireball** [comment]
@SpyderRoll I have a couple problems now:

1. Is it acceptable for me to try reverse engineering iTools?
2. iTools is super shady, and is already flagged as malware on VirusTotal.

**Comment by AjayGhale** [comment]
i think its aceptable, i have done that before but dont remember for what....

could you guys help me? i managed to display the installed apps of the device on a ListBox using 

`iph.GetApplicationList();
            {
                string[] appNames = iph.GetApplicationList().Select(app=>app.Name).ToArray();

                string[] appVersion = iph.GetApplicationList().Select(app => app.Version).ToArray();

                this.listBox1.Items.AddRange(appNames);
            }`
but the apps are displayed like simple text, i just want to display the apps like on iTools for example, the app icon and below the app name is there any way one of you guys help me to do that?

**Comment by AjayGhale** [comment]
also a way to display the filesystem on listTreeView i guess, using AFC im really rusty on C# now..
and im really intrested on making this app just for learning i would help as far as i can in this project thanks in advance

**Comment by 0xFireball** [comment]
@AjayGhale please move this to another issue. It's not relevant to the discussion here.
